zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -158,46 +187,176 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
+                        // not affect the random linear combination.
+                        tx_nonce_bytes
+                            .iter()
+                            .take($n)
+                            .map(Expr::expr)
+                            .collect::<Vec<_>>()
+                            .as_slice(),
+                    );
+                    // nonce / nonce-prefix
+                    input_exprs.push(select::expr(
+                        tx_nonce_is_zero.expr(),
+                        128.expr(),
+                        select::expr(
+                            tx_nonce_lt_128.expr(),
+                            tx_nonce.expr(),
+                            128.expr() + tx_nonce_byte_size.byte_size(),
+                        ),
+                    ));
+                    // address
+                    input_exprs.extend_from_slice(
+                        tx_caller_address_bytes
+                            .iter()
+                            .map(Expr::expr)
+                            .collect::<Vec<_>>()
+                            .as_slice(),
+                    );
+                    // addr-prefix
+                    input_exprs.push(148.expr());
+                    // prefix
+                    input_exprs.push(select::expr(
+                        tx_nonce_lt_128.expr(),
+                        214.expr(),
+                        214.expr() + tx_nonce_byte_size.byte_size(),
+                    ));
+                    let input_exprs: [Expression<F>; MIN_RLP_LEN_CONTRACT_CREATION_INPUT + $n] =
+                        input_exprs.try_into().unwrap();
+                    let input_rlc = cb.keccak_rlc(input_exprs);
+
+                    // 4. keccak table lookup
+                    cb.keccak_table_lookup(input_rlc, input_len.expr(), output_rlc.expr());
+                });
+            };
+        }
+
+        // Cover various cases of keccak lookup guarded by the number of bytes used to
+        // represent transaction's nonce.
+        //
+        // The tx nonce in the RLP-encoding could be n ∈ [0, 8] bytes. Depending on the
+        // byte size of tx nonce, we take `n` byte from the 8-bytes nonce array.
+        keccak_lookup!(0, tx_nonce_lt_128.expr());
+        keccak_lookup!(
+            1,
+            not::expr(tx_nonce_lt_128.expr()),
+            tx_nonce_byte_size_cmp.value_equals(1usize)
+        );
+        keccak_lookup!(2, tx_nonce_byte_size_cmp.value_equals(2usize));
+        keccak_lookup!(3, tx_nonce_byte_size_cmp.value_equals(3usize));
+        keccak_lookup!(4, tx_nonce_byte_size_cmp.value_equals(4usize));
+        keccak_lookup!(5, tx_nonce_byte_size_cmp.value_equals(5usize));
+        keccak_lookup!(6, tx_nonce_byte_size_cmp.value_equals(6usize));
+        keccak_lookup!(7, tx_nonce_byte_size_cmp.value_equals(7usize));
+        keccak_lookup!(8, tx_nonce_byte_size_cmp.value_equals(8usize));
```

Comment:
With the RLP gadget, I believe it will be possible to remove this. Before merging this PR, I'll make it a point to create that issue and define the scope.

Reply:
@CPerezz I've ported over @z2trillion 's work with tests, which includes the gadget to calculate `CREATE` address is here:
https://github.com/scroll-tech/zkevm-circuits/blob/feat/contract-create-rlp/zkevm-circuits/src/evm_circuit/util/math_gadget/rlp.rs#L182

I'd suggest I can do the refactoring as part of a separate PR (to make the reviewing process more convenient). But if you think it's better to be wrapped up in this very PR, happy to make the changes :)

Reply:
Sure if we file a `tech-debt` PR before merging with all of the things like this.

Reply:
This part of multiple condition guarded keccak lookups has been replaced by a single keccak lookup, where the input rlc and input length is calculated from the `ContractCreate` gadget that we merged earlier.